added set-function and unset-function commands

nanobowers opened the pull request:
I've added the `set-function` and `unset-function` commands and tested them locally using sh type shells and fish.  I have not found documentation on how to export a function in fish, it is unclear if that is possible.

I am not a fish user, but one thing I have noticed is that fish syntax is significantly different from sh type shells and not compatible.  Thus any functions that use command-substitution or function arguments will be need to be different.  To support both using the same set-function command may involve duplicate definitions, e.g.:
```tcl
if { [module-info shelltype sh] } then {
    set-function mycd {eval `mycd.sh $1`}
} elseif { [module-info shelltype fish] } then {
    set-function mycd {eval (mycd.sh $argv[1])}
}
```
In this case, it might be useful to define helper functions so that it is more seamless, e.g:

```tcl
## Shell helper function
## Accepts either a number or "all", for all arguments.
proc shell-arg {argnum} {
    if { [module-info shelltype fish] } then {
	if { $argnum == "all" } then {
	    return "\$argv"
	} else {
	    return "\$argv\[${argnum}\]"
	}
    } else {
	if { $argnum == "all" } then {
	    return "\"\$@\""
	} else {
	    return "\$${argnum}"
	}
    }
}

## Shell helper function
## Wraps a command with the appropriate command substitution
## delimiters.  Either:
##  (cmd)  ==> for fish
##  `cmd`  ==> for all other shells
proc shell-commandsub {args} {
    if { [module-info shelltype fish] } then {
	return "([join $args])"
    }
    return "\`[join $args]\`"
}
    
set-function mycd "eval [shell-commandsub mycd.sh [shell-arg 1]]"
```

I'm not sure if it would be best to let users handle this, or add these `shell-arg` and `shell-commandsub` convenience procedures to the modulecmd.tcl - **please advise**.  Supporting these helper procs within modules may open up the door to additional feature requests.

No tests for set-function / unset-function, will wait on feedback.

xdelaruelle commented on modulecmd.tcl.in:
```diff
@@ -4036,6 +4063,33 @@ proc renderSettings {} {
+            switch -- $::g_shellType {
+               {sh} {
+                  #set val [charEscaped $val]
+		  puts stdout "$funcname () { $val ;} ; export $funcname;"
```
I wonder if a function is exported the same way over all sh-kind shells (sh, bash, ksh & zsh). Especially `bash` has a `-f` optional argument to `export` command for function. Isn't it useful here ?

nanobowers commented:
I think i've fixed the issues you gave, added documentation and tests for set-function/unset-function.
Some comments/questions:
there is are some tests related to the `avail` command that are now failing, due to the fact that I added two new modulefiles called function/1.0 and function/2.0.  What is the appropriate way to fix this test?

Also, is there a simple way to quickly rerun one test at a time?  Rerunning `make test` is quite lengthy, but I could not find any documentation nor was it obvious to me from digging in the code.

nanobowers commented:
> > I'm not sure if it would be best to let users handle this, or add these shell-arg and shell-commandsub convenience procedures to the modulecmd.tcl - please advise. Supporting these helper procs within modules may open up the door to additional feature requests.
> 
> These `shell-arg` and `shell-commandsub` procedures you propose are very nice ideas. However I would prefer in a first time to let users handle this. Will think on these helper procedures to normalize shell code in the future, and what it imply globally.

Regarding this, is there an easy way to make a users tcl procs accessible to modules?  Ideally I would want to define these procs in one file so that I could use the functions in any modulefile.

xdelaruelle commented:
> I think i've fixed the issues you gave, added documentation and tests for set-function/unset-function.

Seems great! Could you please rebase your fork on the master branch, so only the commits you made for this change will appear here.

> Some comments/questions:
> there is are some tests related to the `avail` command that are now failing, due to the fact that I added two new modulefiles called function/1.0 and function/2.0. What is the appropriate way to fix this test?

Run the testsuite in a verbose mode (`export RUNTESTFLAGS=-v`), then set new answers obtained from these failing tests as expected results. I can handle that part ig you allow me to update your pull request branch [1]

> Also, is there a simple way to quickly rerun one test at a time? Rerunning `make test` is quite lengthy, but I could not find any documentation nor was it obvious to me from digging in the code.

I have just added a small utility called `mt` in the contrib directory. Launching it from the root directory of the repository and it will run the test suite. This tool can run specific parts of the suite whose reference code are passed to it as argument, for instance `contrib/mt 50/{390,391}`

[1] https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

xdelaruelle commented:
> > > I'm not sure if it would be best to let users handle this, or add these shell-arg and shell-commandsub convenience procedures to the modulecmd.tcl - please advise. Supporting these helper procs within modules may open up the door to additional feature requests.
> > 
> > 
> > These `shell-arg` and `shell-commandsub` procedures you propose are very nice ideas. However I would prefer in a first time to let users handle this. Will think on these helper procedures to normalize shell code in the future, and what it imply globally.
> 
> Regarding this, is there an easy way to make a users tcl procs accessible to modules? Ideally I would want to define these procs in one file so that I could use the functions in any modulefile.

Not yet available but I would like to introduce a `module-proc` modulefile command to achieve that like tested here [1]. Currently these procedures can be integrated to the modulefile visibility by hacking the `initModfileModeAliases` procedure.

[1] https://github.com/bertwesarg/modules-tcl/commit/907ce965d5455939a37f6a0fb54c49d926a81808

nanobowers commented:
I think I got it right this time and did a rebase plus squash my commits down.  The contrib/mt was a real time saver, thanks!  I enabled edits-by-maintainer, so please feel free to make any changes.

xdelaruelle commented:
Perfect! Just in time for 4.2 release
